### PR TITLE
scripts: shell: bt_nus_shell: Convert to python3

### PR DESCRIPTION
**Pull request by nordic-krch**
Converted script to python 3. Kept python2 version as `scripts/shell/bt_nus_shell_py2.py` for now.

Updated documentation.

### CODE REVIEW

**Comment by qung** [comment]
Hi @nordic-krch should we clear state python3 instead of python?

File: scripts/shell/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+pc-ble-driver-py>=0.15.0
```
Comment (rafpelc) [inline scripts/shell/requirements.txt:1]: i've tested only on v.0.14.2 so can't be sure if this works from my perspective
if scripts works with v0.15.0 ignore this comment :)

File: scripts/shell/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
...
+nrfutil>=6.0.0
```
Comment (rafpelc) [inline scripts/shell/requirements.txt:2]: i've tested only on v.6.1.0 so can't be sure if this works from my perspective
if scripts works with v6.0.0 ignore this comment :)

**Review by rafpelc — APPROVED** [review]

File: scripts/shell/requirements.txt (hunk 1 above)
Comment (nordic-krch) [reply scripts/shell/requirements.txt:1]: i just added what i tested with.

**Comment by nordic-krch** [comment]
> Hi @nordic-krch should we clear state python3 instead of python?

@qung done.

**Review by qung — APPROVED** [review]

**Comment by nordic-krch** [comment]
Compliance fail is caused by pylinting python2 file (` E0001: Missing parentheses in call to 'print'`)

**Review by FrancescoSer — APPROVED** [review]

**Comment by nordic-krch** [comment]
@rlubos can you merge it. Red compliance check will not be fixed:
> Compliance fail is caused by pylinting python2 file ( E0001: Missing parentheses in call to 'print')

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@nordic-krch there's also this:
`Missing newline at end of 'scripts/shell/requirements.txt'. Check your text editor settings.`
can you fix it?

**Comment by nordic-krch** [comment]
@carlescufi done.

**Comment by carlescufi** [comment]
> @carlescufi done.

There are still some issues with compliance that can be fixed, could you take a look please?

https://github.com/nrfconnect/sdk-nrf/pull/3308/checks?check_run_id=1411237155

**Comment by nordic-krch** [comment]
@carlescufi yes, sorry for that, must have messed something with the commit. Now compliance looks better

**Comment by tejlmand** [comment]
> Kept python2 version as scripts/shell/bt_nus_shell_py2.py for now.

Why keep the Python2 version ?
We clearly state Python 3.6 is minimum version for NCS:
https://github.com/nrfconnect/sdk-nrf/blob/dc247d2f2239b5e021a1bc589e7827158283cc56/scripts/tools-versions-minimum.txt#L1

and in the docs:
http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/gs_recommended_versions.html

File: scripts/shell/bt_nus_shell.py [hunk 3]
```diff
@@ -1,84 +1,87 @@
-#!/usr/bin/env python
 # Copyright (c) 2019 Nordic Semiconductor ASA
```
Comment (tejlmand) [inline scripts/shell/bt_nus_shell.py:1]: Please add:
```suggestion
#!/usr/bin/env python3
# Copyright (c) 2019 Nordic Semiconductor ASA
```
```

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: scripts/shell/bt_nus_shell.py (hunk 3 above)
Comment (nordic-krch) [reply scripts/shell/bt_nus_shell.py]: done

**Comment by nordic-krch** [comment]
@tejlmand removed python2 version.

**Review by tejlmand — APPROVED** [review]

File: scripts/shell/bt_nus_shell.py [hunk 4]
```diff
@@ -1,84 +1,88 @@
...
+from pc_ble_driver_py.ble_driver import ATT_MTU_DEFAULT
+from pc_ble_driver_py.ble_adapter import BLEAdapter, BLEAdapterObserver, EvtSync
+
+
+CFG_TAG = 1
 
 parser = argparse.ArgumentParser(description='BLE serial daemon.')
 parser.add_argument('--com', help='COM port name (e.g. COM110).',
-                    nargs='*', required=False, default="")
+                    nargs='*', required=True, default="")
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:29]: Switching `--com` from `required=False` to `required=True` is a user visible change that is not related to Python 3 which is not explained in the commit log. Why is it in this commit? Why is the change being done?

File: scripts/shell/bt_nus_shell.py [hunk 5]
```diff
@@ -1,84 +1,88 @@
...
                     nargs=1, required=False, default="")
 parser.add_argument('--name', help='Device name (advertising name).',
-                    nargs='*', required=True, default=None)
-args = parser.parse_args()
-
-def ble_driver_init(conn_ic_id):
-    global BLEDriver, Flasher, DfuTransportBle
-    from pc_ble_driver_py import config
-    config.__conn_ic_id__ = conn_ic_id
+                    nargs='*', required=False, default=None)
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:35]: Switching `--name` from `required=True` to `required=False` is a user visible change not related to Python 3 which is not explained in the commit log. Why is it in this commit? Why is the change being done?

File: scripts/shell/bt_nus_shell.py [hunk 6]
```diff
@@ -1,84 +1,88 @@
...
+parser.add_argument('--port', help='Local port to use.',
+                    required=False, default=8889)
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:37]: Adding a new `--port` option is a user visible change not related to Python 3. I can guess why it's being done, but why is it in this commit?

File: scripts/shell/bt_nus_shell.py [hunk 7]
```diff
@@ -1,84 +1,88 @@
-#!/usr/bin/env python
-# Copyright (c) 2019 Nordic Semiconductor ASA
+#!/usr/bin/env python3
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:1]: There's nothing wrong with this line, but the commit message still says:

> renamed deprecated python2 version to bt_nus_shell_py2.py

Which is not true anymore

File: scripts/shell/bt_nus_shell.py [hunk 8]
```diff
@@ -1,84 +1,88 @@
...
-from nordicsemi.dfu.dfu_transport_ble import DFUAdapter
-from pc_ble_driver_py.exceptions      import NordicSemiException, IllegalStateException
-from pc_ble_driver_py.ble_driver      import config,Flasher,BLEDriver, BLEDriverObserver, BLEEnableParams, BLEUUIDBase, BLEUUID, BLEAdvData, BLEGapConnParams, NordicSemiException
-from pc_ble_driver_py.ble_driver      import ATT_MTU_DEFAULT
-from pc_ble_driver_py.ble_adapter     import BLEAdapter, BLEAdapterObserver, EvtSync
+HOST = ''   # Symbolic name meaning all available interfaces
+PORT =  int(args.port)
 
-global nrf_sd_ble_api_ver
+# global nrf_sd_ble_api_ver
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:44]: ??

File: scripts/shell/bt_nus_shell.py [hunk 9]
```diff
@@ -1,84 +1,88 @@
...
 nrf_sd_ble_api_ver = config.sd_api_ver_get()
-logger  = logging.getLogger(__name__)
+logger = logging.getLogger(__name__)
 
-class BLE_Serial(BLEDriverObserver, BLEAdapterObserver):
-    BASE_UUID   = BLEUUIDBase([0x6E, 0x40, 0x00, 0x00, 0xB5, 0xA3, 0xF3, 0x93,
+
+class BleSerial(BLEDriverObserver, BLEAdapterObserver):
+    BASE_UUID = BLEUUIDBase([0x6E, 0x40, 0x00, 0x00, 0xB5, 0xA3, 0xF3, 0x93,
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:50]: Unrelated changes like this make it harder to see the parts of this commit that are related to the python2 -> python3 conversion.

"Unrelated" because the BLE_Serial -> BleSerial rename and the whitespace change in initializing BASE_UUID are not required for converting to Python 3.

File: scripts/shell/bt_nus_shell.py [hunk 10]
```diff
@@ -1,84 +1,88 @@
...
-        self.evt_sync           = EvtSync(['connected', 'disconnected'])
+    RX_UUID = BLEUUID(0x0003, BASE_UUID)
+    TX_UUID = BLEUUID(0x0002, BASE_UUID)
+    LOCAL_ATT_MTU = 23
+
+    def __init__(self, com_port, periph_name, s_conn):
+        driver = BLEDriver(serial_port    = com_port,
+                           baud_rate      = 1000000)
+        adapter = BLEAdapter(driver)
+        self.evt_sync = EvtSync(['connected', 'disconnected'])
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:61]: These changes are also unrelated to conversion to python 3 as far as I can tell.

They're not fully PEP 8 compliant either -- you would say `serial_port=com_port` and `baud_rate=1000000` (with no extra whitespace) in PEP 8 style.

File: scripts/shell/bt_nus_shell.py [hunk 11]
```diff
@@ -1,84 +1,88 @@
...
         self.target_device_name = periph_name
         self.target_device_addr = 0
-        self.conn_handle        = None
-        self.adapter            = adapter
-        self.notifications_q    = Queue.Queue()
+        self.conn_handle = None
+        self.adapter = adapter
+        self.notifications_q = queue.Queue()
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:66]: Only the last line in this part of the diff is relevant to python 3.

File: scripts/shell/bt_nus_shell.py [hunk 12]
```diff
@@ -1,84 +1,88 @@
...
 
-        self.adapter.driver.ble_enable(ble_enable_params)
-        self.adapter.driver.ble_vs_uuid_add(BLE_Serial.BASE_UUID)
+        gatt_cfg = BLEConfigConnGatt()
+        gatt_cfg.att_mtu = self.adapter.default_mtu
+        gatt_cfg.tag = CFG_TAG
+        self.adapter.driver.ble_cfg_set(BLEConfig.conn_gatt, gatt_cfg)
+
+        self.adapter.driver.ble_enable()
+        self.adapter.driver.ble_vs_uuid_add(BleSerial.BASE_UUID)
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:81]: This seems to be replacing the `BLEEnableParams` object with a new `BLEConfigConnGatt` object. Is that because the underlying API changed when going to python 3? Or is it unrelated to python 3?

File: scripts/shell/bt_nus_shell.py [hunk 13]
```diff
@@ -1,84 +1,88 @@
...
 
         self.connect()
 
     def connect(self):
-        # self.target_device_name = target_device_name
-        # self.target_device_addr = target_device_addr
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:81]: Unrelated to python 3, but if you're removing commented out lines here, why is one being added on line 44? (`# global nrf_sd_ble_api_ver`)

File: scripts/shell/bt_nus_shell.py [hunk 14]
```diff
@@ -91,14 +95,14 @@ def connect(self):
         if nrf_sd_ble_api_ver >= 3:
             if DFUAdapter.LOCAL_ATT_MTU > ATT_MTU_DEFAULT:
                 logger.info('BLE: Enabling longer ATT MTUs')
-                self.att_mtu = self.adapter.att_mtu_exchange(self.conn_handle)
+                self.att_mtu = self.adapter.att_mtu_exchange(self.conn_handle, BleSerial.LOCAL_ATT_MTU)
```
Comment (mbolivar-nordic) [inline scripts/shell/bt_nus_shell.py:98]: How is this Python 3 related?

This PR seems to be mixing up three things:

1. cosmetics like class renames and whitespace changes
2. language changes required for converting to python 3
3. API changes in pc_ble_driver_py

I'm not sure if it's worth it for me to untangle this, so I will stop here and leave it up to others to merge this PR.

**Review by tejlmand — CHANGES_REQUESTED** [review]
Please address comments from @mbolivar-nordic .
Either by improving, or describe why not.

**Comment by nordic-krch** [comment]
@mbolivar-nordic I've got updated version from verification team and applied it to remove python2 dependency.
In fact, it contains more than just aligning to python3. I've rephrased the commit message. I will not have time to split commit into changes related to python3 and other changes.

File: scripts/shell/bt_nus_shell.py (hunk 10 above)
Comment (nordic-krch) [reply scripts/shell/bt_nus_shell.py]: whitespace removed

File: scripts/shell/bt_nus_shell.py (hunk 8 above)
Comment (nordic-krch) [reply scripts/shell/bt_nus_shell.py]: deleted

**Comment by mbolivar-nordic** [comment]
@nordic-krch the documentation patches describing the parameters don't seem to be in sync with the code changes.

**Review by tejlmand — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
> @nordic-krch the documentation patches describing the parameters don't seem to be in sync with the code changes.

Can a docs maintainer comment on this please?

**Comment by FrancescoSer** [comment]
> > @nordic-krch the documentation patches describing the parameters don't seem to be in sync with the code changes.
> 
> Can a docs maintainer comment on this please?

Hi there! Could you please be more specific? Also tagging @nordic-krch

**Comment by mbolivar-nordic** [comment]
I'm basically asking if this is a blocker. If so I think @nordic-krch should update the docs. One example where they are inconsistent is that the docs say these are required:

```
* ``family`` - the chip family of the development kit, for example, nRF52
* ``name`` - the advertising name of the device with the NUS shell
```

But in fact  both of them are optional. The `name` argument used to be required, but is made optional by this PR. The `family` argument was already not required.

I think the docs should be re-checked against the latest code changes.

**Comment by nordic-krch** [comment]
@mbolivar-nordic I've updated documentation to match actual script options.

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by joerchan** [comment]
@nordic-krch Please rebase

**Comment by nordic-krch** [comment]
@joerchan done
